packages/utils-debugging-protocol-common/src/debugging-protocol-connector.ts

```diff
@@ -302,7 +307,19 @@ export class Connector implements IConnector {
         }
     }
 
-    private async getResponseBody(cdpResponse): Promise<{ content: string, rawContent: Buffer, rawResponse(): Promise<Buffer> }> {
+    /** Wait until the given `requestId` request has loaded all the content. */
+    // TODO: remove `any` from return type
+    private waitForContentLoaded(requestId: string): Promise<any> {
+        if (this._finishedRequests.has(requestId)) {
+            return Promise.resolve();
```

Comment:
Since we add items to `_finishedRequests` before they're resolved, couldn't this incorrectly cause the calling code to think the request is complete when it's still pending? Perhaps the `Promise` created on line 317 should be included in the `Map` and returned here instead?

Reply:
I think the problem is the name of the variable. `_finishedRequests` isn't the best one but couldn't find a better one. Let's see if I can explain how this works (it's a bit circular) and we can come out with a better one:

**Option 1:** Request the body after the full response is available:

1. the `onLoadingFinished` (which is fired only when the full response of a request is available) event handler adds an item to the `_finishedRequests` whose key is the `requestId`.
1. `waitingForContentLoaded` is called from `getResponseBody`. It checks if `_finishedRequests` has an item for that `requestId`. It does so it knows the content is available and can actually request it via `Network.getResponseBody`

**Option 2:** Request the body before the full response is available:

1. `waitingForContentLoaded` is called from `getResponseBody`. `_finishedRequests` doesn't have an item with the right `requestId` so it adds a promise to the value to be called when it is available.
1. `onLoadingFinished` receives the vent for the previous `requestId`. Because the item is already there, it calls `resolve` so the execution can continue in the other side

Let me know if this makes more sense. If you have a better idea on how to do it let me know. This is the first thing I could come up (and we are using a similar pattern for something similar here too). Happy to document this also in the code.